### PR TITLE
8209595: MonitorVmStartTerminate.java timed out

### PR DESCRIPTION
**Pull request by kevinjwalls**
From studying test failures, it looks like the way the test identifies its related processes is failing.
It checks the mainArgs of a process by attaching, and looks like it occasionally misses getting a valid match.  The hasMainArgs method ignores exceptions as it is expecting some exceptions: it is going to test unrelated java process which happen to start.

It should retry this main args check on failure, but not too many times to be a burden on other valid unrelated processes, and should also log the PIDs that have an issue so we can see if this is part of any future failure.

Other small logging changes so we can see more easily the progress through the test.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8209595](https://bugs.openjdk.org/browse/JDK-8209595): MonitorVmStartTerminate.java timed out (**Bug** - P4)

### Reviewers
 * [Serguei Spitsyn](https://openjdk.org/census#sspitsyn) (@sspitsyn - **Reviewer**) ⚠️ Review applies to [8501c23c](https://git.openjdk.org/jdk/pull/16077/files/8501c23c29dad0b6774344ddedef33bf81f38bb1)
 * [Chris Plummer](https://openjdk.org/census#cjplummer) (@plummercj - **Reviewer**)
 * [Leonid Mesnik](https://openjdk.org/census#lmesnik) (@lmesnik - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16077/head:pull/16077` \
`$ git checkout pull/16077`

Update a local copy of the PR: \
`$ git checkout pull/16077` \
`$ git pull https://git.openjdk.org/jdk.git pull/16077/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16077`

View PR using the GUI difftool: \
`$ git pr show -t 16077`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16077.diff">https://git.openjdk.org/jdk/pull/16077.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16077#issuecomment-1752728206)

### CODE REVIEW

File: test/jdk/sun/jvmstat/monitor/MonitoredVm/MonitorVmStartTerminate.java [hunk 1]
```diff
@@ -166,23 +168,35 @@ private void releaseTerminated(Set<Integer> ids) {
...
         private boolean hasMainArgs(Integer id, String args) {
-            try {
-                VmIdentifier vmid = new VmIdentifier("//" + id.intValue());
-                MonitoredVm target = host.getMonitoredVm(vmid);
-                String monitoredArgs = MonitoredVmUtil.mainArgs(target);
-                if (monitoredArgs != null && monitoredArgs.contains(args)) {
-                    return true;
+            // As we have seen test timeouts due to missing a notification,
+            // we should retry this attempt to check arguments for a match.
+            for (int i=0; i<ARGS_ATTEMPTS; i++) {
```
Comment (sspitsyn) [inline test/jdk/sun/jvmstat/monitor/MonitoredVm/MonitorVmStartTerminate.java:183]: Nit: Need spaces around `=` and `<`.

File: test/jdk/sun/jvmstat/monitor/MonitoredVm/MonitorVmStartTerminate.java [hunk 2]
```diff
@@ -166,23 +168,35 @@ private void releaseTerminated(Set<Integer> ids) {
...
+                        return false;
+                    }
+                } catch (URISyntaxException | MonitorException e) {
+                    // Pocess probably not running or not ours, e.g.
+                    // sun.jvmstat.monitor.MonitorException: Could not attach to PID
+                    System.out.println("hasMainArgs(" + id + "): " + e);
                 }
-            } catch (URISyntaxException | MonitorException e) {
-                // ok. process probably not running
+                takeNap();
```
Comment (sspitsyn) [inline test/jdk/sun/jvmstat/monitor/MonitoredVm/MonitorVmStartTerminate.java:199]: Nit: I'm thinking if moving this after line 197 would be more clear.

**Review by sspitsyn — APPROVED** [review]
Looks good.
Placed a couple of nits.
Thanks,
Serguei

File: test/jdk/sun/jvmstat/monitor/MonitoredVm/MonitorVmStartTerminate.java [hunk 3]
```diff
@@ -166,23 +168,35 @@ private void releaseTerminated(Set<Integer> ids) {
...
-                String monitoredArgs = MonitoredVmUtil.mainArgs(target);
-                if (monitoredArgs != null && monitoredArgs.contains(args)) {
-                    return true;
+            // As we have seen test timeouts due to missing a notification,
+            // we should retry this attempt to check arguments for a match.
+            for (int i=0; i<ARGS_ATTEMPTS; i++) {
+                try {
+                    VmIdentifier vmid = new VmIdentifier("//" + id.intValue());
+                    MonitoredVm target = host.getMonitoredVm(vmid);
+                        String monitoredArgs = MonitoredVmUtil.mainArgs(target);
```
Comment (plummercj) [inline test/jdk/sun/jvmstat/monitor/MonitoredVm/MonitorVmStartTerminate.java:187]: Fix indent.

File: test/jdk/sun/jvmstat/monitor/MonitoredVm/MonitorVmStartTerminate.java [hunk 4]
```diff
@@ -166,23 +168,35 @@ private void releaseTerminated(Set<Integer> ids) {
...
+                    if (monitoredArgs != null && monitoredArgs.contains(args)) {
+                        System.out.println("hasMainArgs(" + id + "): yes");
+                        return true;
+                    } else {
+                        return false;
+                    }
+                } catch (URISyntaxException | MonitorException e) {
+                    // Pocess probably not running or not ours, e.g.
```
Comment (plummercj) [inline test/jdk/sun/jvmstat/monitor/MonitoredVm/MonitorVmStartTerminate.java:195]: Pocess -> Process

**Review by plummercj — APPROVED** [review]

**Comment by kevinjwalls** [comment]
Eventually I did reproduce a further failure with these test changes, where hasMainArgs is the issue.
For a valid test pid, we got the main args, and they did not match, but we know they SHOULD have due to the additional logging.

MonitoredVmUtil.mainArgs(target) can return "Unknown" or null, so we need to handle this, and not presume that the PID is NOT a test process.  We should retry the main args fetch if (monitoredArgs == null || monitoredArgs.equals("Unknown"))

Also, takeNap and the 100ms delay:
This can thrash and just fill logs with 10,000 lines of messages when failing.
Maybe it was short to reduce latency, but that does not seem criticial.  Make it longer.

**Review by lmesnik — APPROVED** [review]

**Review by plummercj — APPROVED** [review]

**Pull request closed without merging**